File: ch_2/2.3/2.3.1-LongestPrefix/prefix.py
"""
ID: alantao
LANG: PYTHON3
TASK: prefix
"""

letterHash = {}
string = ""
maxIndex = 0

def solve(index: int, level):
  global string, maxIndex
  cur = string[index]
  avaliable = True
  if cur in level:
    if not solve(index + 1, level[cur]):
      avaliable = False
    
  if None in level:
    temp = solve(index, letterHash)
    if not avaliable:
      avaliable = temp
  
  if cur not in level and None not in level:
    return False

  if avaliable:
    maxIndex = max(maxIndex, index)
  return avaliable


def main(inputFile: str, outputFile: str):
  global string
  prefixInput = open(inputFile, 'r')
  prefixOutput = open(outputFile, 'w')

  arr = []
  cur = prefixInput.readline().strip().split()

  while cur != ["."]:
    arr += cur
    cur = prefixInput.readline().strip().split()
  
  for a in arr:
    # add a to letterHash
    level = letterHash
    for char in a:
      if char in level:
        level = level[char]
      else:
        level[char] = {}
        level = level[char]
    level[None] = {}

  # print(letterHash)
  cur = prefixInput.readline().strip()

  while len(cur) != 0:
    string += cur
    cur = prefixInput.readline().strip()

  string += '0'

  prefixInput.close()

  solve(0, letterHash)

  prefixOutput.write(str(maxIndex))
  prefixOutput.write("\n")

  prefixOutput.close()


main("prefix.in", "prefix.out")

File: ch_2/2.3/2.3.1-LongestPrefix/prefix.cpp
/*
ID: alantao
TASK: prefix
LANG: C++14
*/

#include <iostream>
#include <vector>
#include <string>
using namespace std;

vector<string> choices;
vector<bool> dp(200001);
string mainStr;
int N;

int solve() {
  for (int i = 0; i < N; i++) {
    if (!dp[i]) {
      continue;
    }

    for (string choice : choices) {
      // compare choice to mainStr
      bool fits = true;
      for (int j = i; j < i + choice.length(); j++) {
        if (j > N) {
          // index out of range
          fits = false;
          break;
        }
        if (choice[j-i] != mainStr[j]) {
          // not the same
          fits = false;
          break;
        }
      }

      if (fits) {
        dp[i + choice.length()] = true;
      }
    }
  }

  // get first true from dp
  for (int i = N; i >= 0; i--) {
    if (dp[i]) {
      // found
      return i;
    }
  }
  return 0;
}

int main() {
  freopen("prefix.in", "r", stdin);
  freopen("prefix.out", "w", stdout);

  string cur;
  cin >> cur;

  // get the choices
  while (cur != ".") {
    choices.push_back(cur);
    // cout << cur << endl;
    cin >> cur;
  }

  // get the string
  while (cin >> cur) {
    mainStr += cur;
  }

  N = mainStr.length();
  dp[0] = true;

  // print
  cout << solve() << endl;

  return 0;
}